New intro to React Native and React (#1535)
facebook/react-native-website: docs/introduction.md

## Functional and Class Syntaxes

You may have noticed the two tabs for “Functional Syntax” and “Class Syntax” in the Snack above. We thought it would be useful for you to have both syntaxes to study, though we will only go into depth with the Hooks syntax. React used a class-based syntax until [React 16.8 introduced a functional syntax called “Hooks”](https://reactjs.org/docs/hooks-intro.html). In March 2019, [React Native 0.58 adopted Hooks as well](https://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059). While Hooks are the future-facing way to write your React-based components, you might be working in a codebase that still uses classes.

With the functional syntax you export a JavaScript function that returns your component(s):

```
export default function YourApp() {
  return (
      <View style={{ flex: 1, justifyContent: "center", alignItems: "center" }}>
        <Text>Hello, world!</Text>
      </View>
  );
}
```

> > The sample code in this documentation uses [JavaScript version ES2015](https://babeljs.io/learn-es2015/) (also known as ES6). Since ES2015 is not yet supported by all browsers, React Native ships with ES2015 support out of the box thanks to [Babel](https://babeljs.io/) (which also handles your JSX!). This means you can use ES2015 without worrying about backwards compatibility.

Reviewer: I'm not sure that browsers supporting ES2015 impacts React Native. 

Should we just get rid of that part and say "The sample code in this documentation uses JavaScript version ES2015 (also known as ES6). React Native ships with ES2015 support out of the box thanks to Babel...

Author: Switched over to "engines" to be technically correct ;)

---
id: introduction
title: Introduction
---

> So many different kinds of people use React Native: from master iOS developers to React JS beginners, to people getting started programming for the first time in their career. These docs were written for all learners, no matter their experience level or background.

Reviewer 2: ```suggestion
Many different kinds of people use React Native: from advanced iOS developers to React JS beginners, to people getting started programming for the first time in their career. These docs were written for all learners, no matter their experience level or background.
```

      Functional Syntax
    </li>
    <li id="classical" class="button-classical" aria-selected="false" role="tab" tabindex="0" aria-controls="classicaltab" onclick="displayTab('syntax', 'classical')">
      Class Syntax
    </li>
  </ul>
</div>

<block class="functional webNote iosNote androidNote" />

```SnackPlayer name=Hello%20World
import React from 'react';
import { Text, View } from 'react-native';

export default function YourApp() {
  return (
      <View style={{ flex: 1, justifyContent: "center", alignItems: "center" }}>
        <Text>Hello, world!</Text>
      </View>
  );
}
```

<block class="classical webNote iosNote androidNote" />

```SnackPlayer name=Hello%20World
import React, { Component } from 'react';
import { Text, View } from 'react-native';

export default class HelloWorldApp extends Component {
  render() {
    return (
      <View style={{ flex: 1, justifyContent: "center", alignItems: "center" }}>
        <Text>Hello, world!</Text>
      </View>
    );
  }
}

```

<block class="classical functional webNote iosNote androidNote" />

> The above is a Snack Player. It’s [a handy tool created by Expo](https://expo.io/tools#snack) to embed and run React Native projects on and share how they render in platforms like Android and iOS. We will be using them a lot to demonstrate code samples throughout these docs. You can play around with sample code provided in the Snack, or you can [set up an environment on your local machine](getting-started) and paste the code into your `App.js` file there.

Reviewer 2: I think the "call to action" to the docs about setting up your machine should be further up, before the "How to use these docs" section.

Something just like, "It's not necessary, but in case you want to setup your local machine before proceeding you can read the dedicated document". 

I feel that setting up one's machine is one of the most complicated parts of the "learning curve" and having the reference here makes it a bit less visible IMHO

Author: Well, that was a bit intentional. Research was showing learners bogging down on environment setup, which is pretty complicated and can be discouraging to folks who just want to get a feel for React Native. The hope is that by moving it past the tutorial mode to a clearly labeled section and using in-browser examples, folks would feel more confident about investing so much time later. But let me have a try!

Reviewer 2: Oh I see!
I think I see your point - and I think it totally makes sense to "move away" the setup from this area.

Reviewer 3: Could we remove the env setup guidance from this paragraph and move the paragraph above this one here (See my comment above)? I believe the flow will work better that way:

* How do use the docs
* You need to know JS
* Play with it right here
* Optionally install a local setup and come back

vs:

* How do use the docs
* You need to know JS
* Optionally install a local setup and maybe come back?
* Play with it right here

The first one gets people to jump in right away, making the local setup a secondary flow for people who just want to get started.

> The above is a Snack Player. It’s [a handy tool created by Expo](https://expo.io/tools#snack) to embed and run React Native projects on and share how they render in platforms like Android and iOS. We will be using them a lot to demonstrate code samples throughout these docs. You can play around with sample code provided in the Snack, or you can [set up an environment on your local machine](http://localhost:3000/react-native/docs/next/getting-started) and paste the code into your `App.js` file there.

Reviewer 4: ```suggestion
The above is a Snack Player. It’s [a handy tool created by Expo](https://expo.io/tools#snack) to embed and run React Native projects on and share how they render in platforms like Android and iOS. We will be using them a lot to demonstrate code samples throughout these docs. You can play around with sample code provided in the Snack, or you can [set up an environment on your local machine](getting-started) and paste the code into your `App.js` file there.
```

## Developer Notes

People with many different developer experiences are learning React Native. You may have experience with a range of technologies, from web to Android to iOS and more. We try to write for developers from all backgrounds. Sometimes we provide explanations specifc to one platform or another like so:

<div class="toggler">
  <span>Developer Notes</span>
  <span role="tablist" id="toggle-devNotes">
    <button role="tab" class="button-webNote" onclick="displayTab('devNotes', 'webNote')">Web</button>
    <button role="tab" class="button-androidNote" onclick="displayTab('devNotes', 'androidNote')">Android</button>
    <button role="tab" class="button-iosNote" onclick="displayTab('devNotes', 'iosNote')">iOS</button>
  </span>
</div>

<block class="webNote classical functional" />

> Web developers may be familiar with this concept.

<block class="androidNote classical functional" />

> Android developers may be familiar with this concept.

<block class="iosNote classical functional" />

> iOS developers may be familiar with this concept.

<block class="webNote androidNote iosNote  classical functional" />

---

> Now that you know how this guide works, it's time to get to know the foundation of React Native: [Native Components](intro-react-native-components.md).

Reviewer 2: ```suggestion
Now that you know how this guide works, it's time to get to know the foundation of React Native: [Native and Core Components](intro-react-native-components.md).
```

---
id: introduction
title: Introduction
---

Many different kinds of people use React Native: from advanced iOS developers to React JS beginners, to people getting started programming for the first time in their career. These docs were written for all learners, no matter their experience level or background.

## How to use these docs

You can start here and read through these docs linearly like a book; or you can read the specific sections you need. Already familiar with React JS? You can skip that section—or read it for a light refresher.

## Prerequisites

To work with React Native, you will need to have an understanding of JavaScript fundamentals. If you’re new to JavaScript or need a refresher, you can [dive in](https://developer.mozilla.org/en-US/docs/Web/JavaScript) or [brush up](https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript) at Mozilla Developer Network.

> While we do our best to assume no prior knowledge of React, iOS, or Android development, these are valuable topics of study for the aspiring React Native developer. Where sensible, we have linked to resources and articles that go more in depth.

> ### Environment Setup

Reviewer 3: I recommend moving this after "Interactive examples". Reading from the top down it wasn't clear to me what the next step was: should I click this link and go somewhere else or continue to scroll down? I think it's best to get people to place with RN asap, and then recommend "oh, if you don't wanna play with it here on the website, set up a local dev env here […] and then continue the guide."

Author: So interestingly enough, it _was_ further down, but someone suggested it should be closer to the top for readers who want to set up their environment before before getting too far along. I tend to agree with you, though. I think the straighter flow is better. Let me see what Dan's feedback says :)

Reviewer 5: I don’t care strongly either way. Trust you to choose this!

Author: I'm going to leave the dev environment stuff in a callout under the interactive examples. One reference should be enough.

Optionally, if you want to setup a local development environment, [you can follow our guide to setting up your environment on your local machine](getting-started). Our you can follow along with the interactive examples used throughout this guide.

> ## Interactive examples

Reviewer 3: Maybe this one could have a bit more woo, like "Get started immediately with interactive examples"?

      Functional Syntax
    </li>
    <li id="classical" class="button-classical" aria-selected="false" role="tab" tabindex="0" aria-controls="classicaltab" onclick="displayTab('syntax', 'classical')">
      Class Syntax
    </li>
  </ul>
</div>

<block class="functional webNote iosNote androidNote" />

```SnackPlayer name=Hello%20World
import React from 'react';
import { Text, View } from 'react-native';

export default function YourApp() {
  return (
      <View style={{ flex: 1, justifyContent: "center", alignItems: "center" }}>
        <Text>Hello, world!</Text>
      </View>
  );
}
```

<block class="classical webNote iosNote androidNote" />

```SnackPlayer name=Hello%20World
import React, { Component } from 'react';
import { Text, View } from 'react-native';

export default class HelloWorldApp extends Component {
  render() {
    return (
      <View style={{ flex: 1, justifyContent: "center", alignItems: "center" }}>
        <Text>Hello, world!</Text>
      </View>
    );
  }
}

```

<block class="classical functional webNote iosNote androidNote" />

> The above is a Snack Player. It’s [a handy tool created by Expo](https://expo.io/tools#snack) to embed and run React Native projects on and share how they render in platforms like Android and iOS. We will be using them a lot to demonstrate code samples throughout these docs. You can play around with sample code provided in the Snack, or you can [set up an environment on your local machine](getting-started) and paste the code into your `App.js` file there.

Reviewer 3: One more thing: I got really confused on the website that "a handy tool created by Expo" was linked to some other page. I stopped reading there, clicked the link and got lost on the Expo website without knowing what to do next. Expo has its logo and name on the player that link to the website, and we mention it on almost every page. I think we could get away without turning this text into a link and just letting people play with it. If they are curious, they'll check out Expo on their own.

Reviewer 3: Oops, I have one more thought: Could this paragraph encourage people to make a specific change? Maybe the example could have an emoji in it that is commented out and removing the comment will show a 🎉 emoji or something? That would show people the power of RN and Snack right away and get them hooked :)

## Function and Class Syntaxes

You may have noticed the two tabs for “Functional Syntax” and “Class Syntax” in the Snack above. We thought it would be useful for you to have both syntaxes to study, though we will only go into depth with function components.

> Classes were the only way to manage state in a React component until [React 16.8 introduced an API called “Hooks” for managing state and more in function components](https://reactjs.org/docs/hooks-intro.html). In March 2019, [React Native 0.58 adopted Hooks as well](https://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059). While Hooks are the future-facing way to write your React-based components, you might be working in a codebase that still uses classes.

Reviewer 3: The version numbers and dates will get outdated very quickly. The React version doesn't really mean anything to most RN devs. Could we instead assume hooks as the default and simply mentioning ("Hooks were introduced in React Native 0.58" with a link to the blog post, without dates added?

I'm looking at this as somebody from 2023 (yes, your docs will still be around by then). Reading "March 2019" would make me feel "wow that's old".

Author: Heh, I concede your point, sir!

Author: "With React, you can make components using either classes or functions. Originally, class components were the only components that could have state. But since the introduction of React's Hooks API, you can add state and more to function components.

[Hooks were introduced in React Native 0.58.](https://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059), and because Hooks are the future-facing way to write your React components, we wrote this introduction using function component examples. Where useful, we also cover class components under a toggle like so:"

With function components, you export a JavaScript function that returns your component(s):

```jsx
export default function YourApp() {
  return (
    <View style={{flex: 1, justifyContent: 'center', alignItems: 'center'}}>
      <Text>Hello, world!</Text>
    </View>
  );
}
```

> > The sample code in this documentation uses [JavaScript version ES2015](https://babeljs.io/learn-es2015/) (also known as ES6). Since ES2015 is not supported by all JavaScript engines, React Native ships with ES2015 support out of the box thanks to [Babel](https://babeljs.io/) (which also handles your JSX!). This means you can use ES2015 without worrying about backwards compatibility.

Reviewer 3: Could you remove this? All engines that RN runs on since 0.59 are using modern JavaScript (ES2015+) and as such Babel is not required to ship modern features, even though the transform is still enabled by default.

(cc @ide – hey, should we do anything about that? We should be able to disable almost all babel transforms in RN by default but we'd need to figure out a migration strategy. Good app size and possibly perf impact to be had.)

Author: Good to know, will do!

> ## Developer Notes

Reviewer 3: I love this. In a follow-up, we should probably make these tabs store their state in a cookie/localStorage so it persists across pages.

Author: That would be a lovely thing. Maybe in a ticket for another push?

> People with many different developer experiences are learning React Native. You may have experience with a range of technologies, from web to Android to iOS and more. We try to write for developers from all backgrounds. Sometimes we provide explanations specifc to one platform or another like so:

Reviewer 3: Maybe "Many people with a different developer experience"?

<div class="toggler">
  <span>Developer Notes</span>
  <span role="tablist" id="toggle-devNotes">
    <button role="tab" class="button-webNote" onclick="displayTab('devNotes', 'webNote')">Web</button>
    <button role="tab" class="button-androidNote" onclick="displayTab('devNotes', 'androidNote')">Android</button>
    <button role="tab" class="button-iosNote" onclick="displayTab('devNotes', 'iosNote')">iOS</button>
  </span>
</div>

<block class="webNote classical functional" />

> Web developers may be familiar with this concept.

<block class="androidNote classical functional" />

> Android developers may be familiar with this concept.

<block class="iosNote classical functional" />

> iOS developers may be familiar with this concept.

<block class="webNote androidNote iosNote  classical functional" />

## Formatting

> Menu paths are written in bold and use carats to navigate submenus. Example: **Android Studio > Preferences**

Reviewer 3: Nit: `carets`.

<div class="toggler">
  <ul role="tablist" class="toggle-syntax">
    <li id="functional" class="button-functional" aria-selected="false" role="tab" tabindex="0" aria-controls="functionaltab" onclick="displayTabs('syntax', 'functional')">
      Function Component Example
    </li>
    <li id="classical" class="button-classical" aria-selected="false" role="tab" tabindex="0" aria-controls="classicaltab" onclick="displayTabs('syntax', 'classical')">
      Class Component Example
    </li>
  </ul>
</div>

<block class="functional syntax" />

```SnackPlayer name=Hello%20World%20Function%20Component
import React from 'react';
import { Text, View } from 'react-native';

export default function YourApp() {
  return (
    <View style={{
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center'
      }}>
      <Text>Hello, world!</Text>
    </View>
  );
}
```

<block class="classical syntax" />

```SnackPlayer name=Hello%20World%20Class%20Component
import React, { Component } from 'react';
import { Text, View } from 'react-native';

> export default class HelloWorldApp extends Component {

Reviewer 5: Nit: this one is `HelloWorldApp` but the other one is `YourApp`. Use the same name?

Author: Sloppy!

## Developer Notes

People with many different developer experiences are learning React Native. You may have experience with a range of technologies, from web to Android to iOS and more. We try to write for developers from all backgrounds. Sometimes we provide explanations specifc to one platform or another like so:

<div class="toggler">
  <span>Developer Notes</span>
  <span role="tablist" class="toggle-devNotes">
    <button role="tab" class="button-webNote" onclick="displayTabs('devNotes', 'webNote')">Web</button>
    <button role="tab" class="button-androidNote" onclick="displayTabs('devNotes', 'androidNote')">Android</button>
    <button role="tab" class="button-iosNote" onclick="displayTabs('devNotes', 'iosNote')">iOS</button>
  </span>
</div>

<block class="webNote devNotes" />

> Web developers may be familiar with this concept.

<block class="androidNote devNotes" />

> Android developers may be familiar with this concept.

<block class="iosNote devNotes" />

> iOS developers may be familiar with this concept.

<block class="endBlock devNotes" />

## Formatting

Menu paths are written in bold and use carats to navigate submenus. Example: **Android Studio > Preferences**

> Inline code

Reviewer 5: I'm not sure what this is referring to. Was this meant to be in backticks? Or is this a header?

![Screen Shot 2020-01-15 at 8 07 34 PM](https://user-images.githubusercontent.com/810438/72467378-b22a5500-37d2-11ea-9b15-c6011ae14a78.png)

Author: What the... that looks like something is broken!
